source/util/move_to_front.h

```diff
@@ -305,6 +314,113 @@ class MoveToFront {
+  // Cache for the last accessed value in the sequence.
+  Val last_accessed_value_ = Val();
+  bool last_accessed_value_valid_ = false;
+};
+
+template <typename Val>
+class MultiMoveToFront {
+ public:
+  // Inserts |value| to sequence with handle |mtf|.
+  // Returns false if |mtf| already has |value|.
+  bool Insert(uint32_t mtf, const Val& value) {
+    if (GetMtf(mtf).Insert(value)) {
+      val_to_mtfs_[value].insert(mtf);
+      return true;
+    }
+    return false;
+  }
+
+  // Removes |value| from sequence with handle |mtf|.
+  // Returns false if |mtf| doesn't have |value|.
+  bool Remove(uint32_t mtf, const Val& value) {
+    if (GetMtf(mtf).Remove(value)) {
+      val_to_mtfs_[value].erase(mtf);
+      return true;
+    }
+    assert(val_to_mtfs_[value].count(mtf) == 0);
+    return false;
+  }
+
+  // Removes |value| from all sequences which have it.
+  void RemoveFromAll(const Val& value) {
+    auto it = val_to_mtfs_.find(value);
+    if (it == val_to_mtfs_.end())
+      return;
+
+    auto& mtfs_containing_value = it->second;
+    for (uint32_t mtf : mtfs_containing_value) {
+      GetMtf(mtf).Remove(value);
+    }
+
+    val_to_mtfs_.erase(value);
+  }
+
+  // Computes rank of |value| in sequence |mtf|.
+  // Returns false if |mtf| doesn't have |value|.
+  bool RankFromValue(uint32_t mtf, const Val& value, uint32_t* rank) {
+    return GetMtf(mtf).RankFromValue(value, rank);
+  }
+
+  // Finds |value| with |rank| in sequence |mtf|.
+  // Returns false if |rank| is out of bounds.
+  bool ValueFromRank(uint32_t mtf, uint32_t rank, Val* value) {
+    return GetMtf(mtf).ValueFromRank(rank, value);
+  }
+
+  // Returns size of |mtf| sequence.
+  uint32_t GetSize(uint32_t mtf) {
+    return GetMtf(mtf).GetSize();
+  }
+
+  // Promotes |value| in all sequences which have it.
+  void Promote(const Val& value) {
+    const auto it = val_to_mtfs_.find(value);
+    if (it == val_to_mtfs_.end())
+      return;
+
+    const auto& mtfs_containing_value = it->second;
+    for (uint32_t mtf : mtfs_containing_value) {
+      GetMtf(mtf).Promote(value);
+    }
+  }
+
+  // Inserts |value| in sequence |mtf| or promotes if it's already there.
+  void InsertOrPromote(uint32_t mtf, const Val& value) {
+    if (!Insert(mtf, value)) {
+      GetMtf(mtf).Promote(value);
+    }
+  }
+
+  // Returns if |mtf| sequence has |value|.
+  bool HasValue(uint32_t mtf, const Val& value) {
+    return GetMtf(mtf).HasValue(value);
+  }
+
+ private:
+  // Returns actual MoveToFront object corresponding to |handle|.
+  // As multiple operations are often performed consecutively for the same
+  // sequence, the last returned value is cached.
+  MoveToFront<Val>& GetMtf(uint32_t handle) {
+    if (!cached_mtf_ || cached_handle_ != handle) {
+      cached_handle_ = handle;
+      cached_mtf_ = &mtfs_[handle];
+    }
+
+    return *cached_mtf_;
+  }
+
+  // Container holding MoveToFront objects. Map key is sequence handle.
+  std::map<uint32_t, MoveToFront<Val>> mtfs_;
```

Comment:
std::map is used to have predictable lifetime for *cached_mtf_, I don't think unordered_map is guaranteed to preserve it.